lmdeploy/serve/openai/reasoning_parser/deepseek_r1_reasoning_parser.py

```diff
@@ -0,0 +1,141 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# modified from https://github.com/vllm-project/vllm/tree/v0.7.3/vllm/entrypoints/openai/reasoning_parsers
+import re
+from typing import Optional, Sequence, Tuple, Union
+
+from lmdeploy.serve.openai.protocol import ChatCompletionRequest, DeltaMessage
+
+from .reasoning_parser import ReasoningParser, ReasoningParserManager
+
+
+@ReasoningParserManager.register_module(name='deepseek-r1')
+class DeepSeekR1ReasoningParser(ReasoningParser):
+    """Reasoning parser for DeepSeek R1 model.
+
+    The DeepSeek R1 model uses <think>...</think> tokens to denote reasoning text. This parser extracts the reasoning
+    content from the model output.
+    """
+
+    def __init__(self, tokenizer: object):
+        super().__init__(tokenizer)
+        self.think_start_token = '<think>'
+        self.think_end_token = '</think>'
+
+        self.reasoning_regex = re.compile(rf'{self.think_start_token}(.*?){self.think_end_token}', re.DOTALL)
+
+        if not self.model_tokenizer:
+            raise ValueError('The model tokenizer must be passed to the ReasoningParser '
+                             'constructor during construction.')
+
+        self.think_start_token_id = self.vocab.get(self.think_start_token)
+        self.think_end_token_id = self.vocab.get(self.think_end_token)
+        if (self.think_start_token_id is None or self.think_end_token_id is None):
+            raise RuntimeError('DeepSeek R1 reasoning parser could not locate think start/end '
+                               'tokens in the tokenizer!')
+
+    def extract_reasoning_content_streaming(
+        self,
+        previous_text: str,
+        current_text: str,
+        delta_text: str,
+        previous_token_ids: Sequence[int],
+        current_token_ids: Sequence[int],
+        delta_token_ids: Sequence[int],
+        **kwargs,
+    ) -> Union[DeltaMessage, None]:
+        """Instance method that should be implemented for extracting reasoning
+        from an incomplete response; for use when handling reasoning calls and
+        streaming.
+
+        Has to be an instance method because  it requires state - the current tokens/diffs, but also the information
+        about what has previously been parsed and extracted (see constructor)
+        """
+        # Skip single special tokens
+        if len(delta_token_ids) == 1 and (delta_token_ids[0] in [self.think_start_token_id, self.think_end_token_id]):
+            return None
+
+        # Check if <think> is present in previous or delta.
+        # Keep compatibility with models that don't generate <think> tokens.
```

Comment:
We should update the `deepseek-r1` chat template. The `message2prompt` should output `<think>` at the end of the prompt